### PR TITLE
Adding default cuda stream priority option

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

### CODE REVIEW

File: dali/util/user_stream.h [hunk 1]
```diff
@@ -42,12 +42,14 @@ class DLL_PUBLIC UserStream {
...
   DLL_PUBLIC cudaStream_t GetStream(const dali::Buffer<GPUBackend> &b) {
     size_t dev = GetDeviceForBuffer(b);
-    std::unique_lock<std::mutex> lock(m_);
+    std::lock_guard<std::mutex> lock(m_);
     auto it = streams_.find(dev);
     if (it != streams_.end()) {
       return it->second;
     } else {
-      CUDA_CALL(cudaStreamCreateWithFlags(&streams_[dev], cudaStreamNonBlocking));
+      constexpr int kDefaultStreamPriority = 0;
```
Comment (Kh4L) [inline dali/util/user_stream.h]: Shouldnt the default priority be an argument of this function?

Comment (jantonguirao) [reply dali/util/user_stream.h]: we could, but after discussing with Janusz we decided to let the default priority for those user streams (used when invoking `as_cpu` for instance)

Comment (Kh4L) [reply dali/util/user_stream.h]: Sure

**Comment by jantonguirao** [comment]
> It would be good to add some check using cudaDeviceGetStreamPriorityRange is provided priority is valid.

@JanuszL Done

**Comment by jantonguirao** [comment]
Build 693802

**Comment by jantonguirao** [comment]
Build 693870

File: dali/pipeline/pipeline.h [hunk 2]
```diff
@@ -333,19 +335,35 @@ class DLL_PUBLIC Pipeline {
     this->bytes_per_sample_hint_ = bytes_per_sample_hint;
     this->set_affinity_ = set_affinity;
     this->max_num_stream_ = max_num_stream;
+    this->default_cuda_stream_priority_ = default_cuda_stream_priority;
     this->prefetch_queue_depth_ = prefetch_queue_depth;
     DALI_ENFORCE(batch_size_ > 0, "Batch size must be greater than 0");
+
+    int lowest_cuda_stream_priority = std::numeric_limits<int>::min();
```
Comment (Kh4L) [inline dali/pipeline/pipeline.h]: Maybe ` int lowest_cuda_stream_priority;` ?

File: dali/pipeline/util/stream_pool.h [hunk 3]
```diff
@@ -53,15 +53,14 @@ class StreamPool {
    * @brief Returns a stream from the pool. If max_size has been exceeded,
    * we hand out previously allocated streams round-robin.
    */
-  cudaStream_t GetStream() {
+  cudaStream_t GetStream(int stream_priority) {
```
Comment (mzient) [inline dali/pipeline/util/stream_pool.h]: This actually can give you a stream with different priority. You could make the `stream_priority` a member of `StreamPool` and then the `GetStream` will always return streams with consistent priority.

**Comment by jantonguirao** [comment]
Build 695118

**Pull request merged**